From 547db4f0c74bc5ebf4855b49b2ecd3713406f368 Mon Sep 17 00:00:00 2001
From: Abylay Ospan <aospan@netup.ru>
Date: Fri, 13 Jan 2017 07:32:04 +0300
Subject: [PATCH] arm: cpu: fix second CPU cannot go to online

force second CPU to TWI/TWE writing 'magic' phy address to lut
without this 'hack' second CPU fail to go online later (in platsmp)

Signed-off-by: Abylay Ospan <aospan@netup.ru>
---
 arch/arm/kernel/head.S | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/arch/arm/kernel/head.S b/arch/arm/kernel/head.S
index 04286fd..efa633a 100644
--- a/arch/arm/kernel/head.S
+++ b/arch/arm/kernel/head.S
@@ -27,6 +27,9 @@
 #include CONFIG_DEBUG_LL_INCLUDE
 #endif
 
+#define SOC_ROM_BASE_PA   0xffff0000
+#define SOC_ROM_LUT_OFF   0x400
+
 /*
  * swapper_pg_dir is the virtual address of the initial page table.
  * We place the page tables 16K below KERNEL_RAM_VADDR.  Therefore, we must
@@ -91,6 +94,15 @@ ENTRY(stext)
 	@ ensure svc mode and all interrupts masked
 	safe_svcmode_maskall r9
 
+	/* force second CPU to TWI/TWE 
+	 * without this 'hack' second CPU fail to go online later
+	*/
+	ldr r1, =SOC_ROM_BASE_PA
+	ldr r0, =0xffff1234
+	ldr r0, =0xffff002c
+	str r0, [r1, #SOC_ROM_LUT_OFF]
+	nop
+
 	mrc	p15, 0, r9, c0, c0		@ get processor id
 	bl	__lookup_processor_type		@ r5=procinfo r9=cpuid
 	movs	r10, r5				@ invalid processor (r5=0)?
-- 
2.7.4

